File: python/Trees/binary_tree_maximum_path_sum.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        """
        Finds the maximum path sum in a binary tree. A path can start and end at any node,
        but must go downwards (traveling only from parent nodes to child nodes).
        Approach:
        - Use DFS to compute the maximum path sum for each node considering two cases:
          1. The path passes through the node and possibly both children (split path).
          2. The path continues from the node to one child (no split, for parent recursion).
        - At each node, update the global result with the maximum split path sum.
        - Return the maximum sum of a path starting from the current node and going down one side.

        Time Complexity: O(N), where N is the number of nodes (each node visited once).
        Space Complexity: O(H), where H is the height of the tree (recursion stack).
        """
        res = [root.val]  # type: ignore 

        # Return max path sum without split (for parent recursion)
        def dfs(root):
            if not root:
                return 0

            leftMax = dfs(root.left)
            rightMax = dfs(root.right)
            # Ignore negative paths (treat as 0)
            leftMax = max(leftMax, 0)
            rightMax = max(rightMax, 0)

            # Compute max path sum with split at current node
            res[0] = max(res[0], root.val + leftMax + rightMax)

            # Return max path sum without split (for parent recursion)
            return root.val + max(leftMax, rightMax)
        
        dfs(root)
        return res[0]

# -------------------
# Helper Functions for Testing
# -------------------
def build_tree_from_list(lst):
    if not lst:
        return None
    nodes = [TreeNode(val) if val is not None else None for val in lst]
    kids = nodes[::-1]
    root = kids.pop()
    for node in nodes:
        if node:
            if kids: node.left = kids.pop()
            if kids: node.right = kids.pop()
    return root

# -------------------
# Test Cases
# -------------------
def test_max_path_sum():
    s = Solution()
    # Test case 1: Example
    root = build_tree_from_list([1,2,3])
    assert s.maxPathSum(root) == 6
    # Test case 2: Negative values
    root = build_tree_from_list([-10,9,20,None,None,15,7])
    assert s.maxPathSum(root) == 42
    # Test case 3: Single node
    root = build_tree_from_list([2])
    assert s.maxPathSum(root) == 2
    # Test case 4: All negative
    root = build_tree_from_list([-3,-2,-1])
    assert s.maxPathSum(root) == -1
    # Test case 5: Mixed values
    root = build_tree_from_list([5,4,8,11,None,13,4,7,2,None,None,None,1])
    assert s.maxPathSum(root) == 48
    print("All test cases passed!")

if __name__ == "__main__":
    test_max_path_sum()

File: C++/Trees/binary_tree_maximum_path_sum.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for a binary tree node.
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;

    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    /**
     * Finds the maximum path sum in a binary tree.
     * 
     * Approach:
     * - Use DFS to compute maximum path sum from each node.
     * - At each node:
     *      → compute left & right max path sums (ignoring negative paths).
     *      → update the global maximum if path through current node is better.
     *      → return max path sum starting from current node and going down one side (to parent).
     *
     * Time Complexity: O(N), where N = number of nodes (each node visited once).
     * Space Complexity: O(H), where H = height of tree (recursion stack).
     */
    int maxPathSum(TreeNode* root) {
        int res = root->val; // Initialize result with root's value
        dfs(root, res);
        return res;
    }

private:
    // Helper DFS function
    int dfs(TreeNode* node, int &res) {
        if (!node) return 0;

        int leftMax = max(0, dfs(node->left, res));  // max path sum on left (ignore if negative)
        int rightMax = max(0, dfs(node->right, res)); // max path sum on right (ignore if negative)

        // Path through current node (can include both left and right)
        res = max(res, node->val + leftMax + rightMax);

        // Return max path sum without split (for parent recursion)
        return node->val + max(leftMax, rightMax);
    }
};

/*
-------------------- LINE-BY-LINE SYNTAX EXPLANATION --------------------

#include <bits/stdc++.h>
    ✅ Includes all standard C++ headers.

using namespace std;
    ✅ Avoids typing std:: everywhere.

struct TreeNode { ... }
    ✅ Defines the binary tree node structure with val, left, and right.

TreeNode() : val(0), left(nullptr), right(nullptr) {}
    ✅ Default constructor initializes val to 0 and children to nullptr.

TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    ✅ Constructor with a value and null children.

TreeNode(int x, TreeNode *left, TreeNode *right)
    ✅ Constructor with a value and specific left & right children.

class Solution { ... }
    ✅ Defines the Solution class.

int maxPathSum(TreeNode* root)
    ✅ Public method to compute maximum path sum.

int res = root->val;
    ✅ Initialize result with root’s value.

dfs(root, res);
    ✅ Call helper function to perform DFS.

return res;
    ✅ Return the computed maximum path sum.

private:
    ✅ Defines private helper functions that can only be called internally.

int dfs(TreeNode* node, int &res)
    ✅ Recursive helper to compute maximum path sum at each node.
    ✅ Passes `res` by reference to keep track of global maximum.

if (!node) return 0;
    ✅ Base case: null node contributes 0.

int leftMax = max(0, dfs(node->left, res));
    ✅ Compute left max path sum, ignoring negative values.

int rightMax = max(0, dfs(node->right, res));
    ✅ Compute right max path sum, ignoring negative values.

res = max(res, node->val + leftMax + rightMax);
    ✅ Update global result if path through current node is better.

return node->val + max(leftMax, rightMax);
    ✅ Return max path sum without split for parent recursion.

*/
